# **Part 1:**
Write, as a text cell, the sequence of steps that you need to perform in order to fullfill your mission.
Recall that your mission consists of providing the following:



*   **Classifier**, a trained sklearn model that they can use to predict 
     potability from the 9 attributes mentioned above
*   **Accuracy**, a floating point value between 0 and 1 indicating how   
     accurate you expect your model to be when used on data not included in 
     the file provided.

For this assignment, limit your choice of classfiers to **k-nearest neighbors, naive Bayes and decision trees.**

You may use the sklearn implementations of all classifiers.

Steps on how I would approach and split the problem

1.   Upload Data from water_potability.xlsx using 
2.   Read data and store into a data frame variable using Pandas 
3.   Read data and store into a data frame variable using Pandas 
     Convert data frame into 2 seperate numpy arrays, where one will be a 1-D 
     array and the other will be a 2-d array
    *   1-D array will represent y_values of if water is safe to drink or not,
        which is taken from the entire column of values from the last column 
        in the data whoch is column index 9
    *   2-D array will represent x_values of water samples and their 9 
        attributes which is taken from the entire column of values from 
        column index 0 to 8

4.  Split both arrays into 3 sets (either  80% train by 20% test or 90% train by 10% test)  where one set will be used for training the model and the other will be used for testing the the accuracy of the models prediction. (x_train, y_train, x_test, y_test)
5. To ensure we have enough data to work with and be able to create the best possible accuracy, I will check every model I train and test with a dataset of 70%_train by 30%_test(added this)dataset of 80%_train by 20%_test and a dataset of 90%_train by 10%_test.

6. Utilize and test as many models and parameters as we can that will help either improve accuracy/MSE and/04 speed for the test data while also seeing which 3 data set splits work best for each model

7. Two main groups of models will be used, classification and regressor identitied. Classification has Knn classification, naive bayes, and decision trees. Regressor has Knn and decision trees.


8. Go through each group and see which model and algorithm produces the best speed and accuracy from comparing the results from the default parameter vs the results with adding changes to the parameters. 

9. Notice the how the results differ from using the 3 different dataset splits, combination of paramters altered, and choose the best algorithm and paramters for each model to compare at the end with best of the others in the group.

10. For both KNN classification and regressor, will test 3 different dataset splits, test default knn, then test a mix and match parameter alterations between weighted vs unweighted knn, changing the value of n, and the knn's different algorithms with sklearn's auto, ball_tree, kd_tree, and brute. 

11. The Naive Bayes has 4 different versions from, Gaussian,	Multinomial, complement, and berouli. Will test defaults on all, will test changes in  alpha/smoothing value for Multinomial, complement, and berouli, and will lastly test changes in binarize value for bernouli.


12. Decision Trees Classification has two main splitting algorithm which are entropy and gini which will be both bested. Each algorithm will be tested with their default paramters, testing best/random splitter, discovering their max_depth in which they prduce the best result, and alter random_state's value. (also Tested the change of max_feature for entropy after noticing the effects it had on the regressor model later on)

13. Decision Trees regressor Each algorithm will be tested with their default paramters, algorithm/criterion that determines how the tree splits, best/random splitter, discovering their max_depth in which they produce the best result. (Discovered recently change in max_features)

14. After changing various of paramater changes for each model, compare results within its model, attempt to discover best result from mix and matching paramters with their set of changes, and then pick the best of that model. 

15. The best model in my decision is taking into consideration accuracy more than speed due to it would be best to give a model/program with better accuracy vs speed due to the nature of the problem which concerns if the water is safe to drink

16. While going through testing various models and their parameters, our goal for every model is to first get ther best and CONSISTENT accuracy/results, the get the best speed for that accuracy, and noting down interesting behaviors/relations.

17. if possible, do extra credit 





Extra credit steps 



# **Part 2.** 
Implement the steps described above as a Python program.

In [ ]:
#upload file/files
from google.colab import files
uploaded = files.upload()

Saving water_potability.xlsx to water_potability.xlsx


In [ ]:
!ls #check file is in colab before starting

sample_data  water_potability.xlsx


In [ ]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import time
from sklearn.metrics import accuracy_score, confusion_matrix,mean_squared_error,mean_absolute_error
from sklearn.metrics import confusion_matrix

from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.naive_bayes import GaussianNB, MultinomialNB, ComplementNB, BernoulliNB

from sklearn.neighbors import KNeighborsRegressor
from sklearn.tree import DecisionTreeRegressor

from sklearn.ensemble import BaggingRegressor
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn import tree
from google.colab import files

water_df = pd.read_excel('water_potability.xlsx')

water_np = water_df.to_numpy() #(2011,10)


#split data into X dataset (water samples and their 9 attributes) 
#and y_values pair (potability found in last column)
X = water_np[:,0:9] #(2011, 9) 2-d array
y = water_np[:,9]   #(2011, 1) 1-d

# 80% by 20% row split = 1608, 403
# 90% by 10% row split = 1809, 202
# random_state=42, popular int random seed (controls randomness)

rs = 42  #Popular/common int value for random_state 0 and 42

#testing 3 sizes of data set splits in all algorithms
x70_train, x30_test, y70_train, y30_test = train_test_split(X, y, test_size=0.3, train_size = .7, random_state=rs)
x80_train, x20_test, y80_train, y20_test = train_test_split(X, y, test_size=0.2, train_size = .8, random_state=rs)
x90_train, x10_test, y90_train, y10_test = train_test_split(X, y, test_size=0.1, train_size = .9, random_state=rs)


# Test out Classifers based on accuracy

1. knn default vs knn mod_paramters

2. Naive Bayes default vs Naive Bayes mod_paramters

3. Decision Tree default vs decision tree mod_paramters

#**Knn-Classification model**

In [ ]:
#default knn 

def default_knnC(xtr, ytr, xte, yte):
  start = time.time()
  model = KNeighborsClassifier().fit(xtr, ytr)

  pred = model.predict(xte)
  accuracy = accuracy_score(yte, pred)
  print(f'Default unweighted knn Accuracy = {accuracy:.4}')

  #default_knn_accuracy = default_knn.score(xte,yte)
  #print("Knn Accuracy ", default_knn_accuracy)

  elapsed_time = time.time() - start
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))

def defaultW_knnC(xtr, ytr, xte, yte):
  start = time.time()
  model = KNeighborsClassifier(weights = 'distance').fit(xtr, ytr)

  pred = model.predict(xte)
  accuracy = accuracy_score(yte, pred)
  print(f'Default Weighted knn Accuracy = {accuracy:.4}')

  #default_knn_accuracy = default_knn.score(xte,yte)
  #print("Knn Accuracy ", default_knn_accuracy)

  elapsed_time = time.time() - start
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))

print("default knn")
default_knnC(x70_train, y70_train, x30_test, y30_test)
default_knnC(x80_train, y80_train, x20_test, y20_test)
default_knnC(x90_train, y90_train, x10_test, y10_test)

print()
print("default weighted knn")
defaultW_knnC(x70_train, y70_train, x30_test, y30_test)
defaultW_knnC(x80_train, y80_train, x20_test, y20_test)
defaultW_knnC(x90_train, y90_train, x10_test, y10_test)



default knn
Default unweighted knn Accuracy = 0.5579
Elapsed time = 0.0279 secs
Default unweighted knn Accuracy = 0.5434
Elapsed time = 0.0181 secs
Default unweighted knn Accuracy = 0.5495
Elapsed time = 0.0115 secs

default weighted knn
Default Weighted knn Accuracy = 0.5546
Elapsed time = 0.0070 secs
Default Weighted knn Accuracy = 0.5285
Elapsed time = 0.0056 secs
Default Weighted knn Accuracy = 0.5545
Elapsed time = 0.0051 secs


Changed N manually and was getting better accuracy results as k increased and 


In [ ]:
#ball_tree weighted and unweighted version with 
#ability to change n 

def ball_tree_knnC(xtr, ytr, xte, yte, n = 15):
  start = time.time()
  model = KNeighborsClassifier(n_neighbors=n, algorithm='ball_tree').fit(xtr, ytr)

  pred = model.predict(xte)
  accuracy = accuracy_score(yte, pred)
  print(f'unweighted KNN balltree Accuracy = {accuracy:.4}')

  #default_knn_accuracy = default_knn.score(xte,yte)
  #print("Knn Accuracy ", default_knn_accuracy)

  elapsed_time = time.time() - start
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))

def Wball_tree_knnC(xtr, ytr, xte, yte, n=15):
  start = time.time()
  model = KNeighborsClassifier(n_neighbors=n, weights = 'distance', algorithm='ball_tree').fit(xtr, ytr)

  pred = model.predict(xte)
  accuracy = accuracy_score(yte, pred)
  print(f'weighted KNN balltree Accuracy = {accuracy:.4}')

  #default_knn_accuracy = default_knn.score(xte,yte)
  #print("Knn Accuracy ", default_knn_accuracy)

  elapsed_time = time.time() - start
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))

for n in range(11, 18, 2):
  print("balltree unweighted with n = ", n)
  ball_tree_knnC(x70_train, y70_train, x30_test, y30_test, n)
  ball_tree_knnC(x80_train, y80_train, x20_test, y20_test, n)
  ball_tree_knnC(x90_train, y90_train, x10_test, y10_test, n)
  print()
  print("balltree weighted with n = ", n)
  Wball_tree_knnC(x70_train, y70_train, x30_test, y30_test, n)
  Wball_tree_knnC(x80_train, y80_train, x20_test, y20_test, n)
  Wball_tree_knnC(x90_train, y90_train, x10_test, y10_test, n)
  print()

balltree unweighted with n =  11
unweighted KNN balltree Accuracy = 0.5762
Elapsed time = 0.0312 secs
unweighted KNN balltree Accuracy = 0.5806
Elapsed time = 0.0188 secs
unweighted KNN balltree Accuracy = 0.5693
Elapsed time = 0.0115 secs

balltree weighted with n =  11
weighted KNN balltree Accuracy = 0.5613
Elapsed time = 0.0070 secs
weighted KNN balltree Accuracy = 0.5533
Elapsed time = 0.0070 secs
weighted KNN balltree Accuracy = 0.5644
Elapsed time = 0.0052 secs

balltree unweighted with n =  13
unweighted KNN balltree Accuracy = 0.5579
Elapsed time = 0.0300 secs
unweighted KNN balltree Accuracy = 0.5682
Elapsed time = 0.0191 secs
unweighted KNN balltree Accuracy = 0.5594
Elapsed time = 0.0112 secs

balltree weighted with n =  13
weighted KNN balltree Accuracy = 0.5629
Elapsed time = 0.0097 secs
weighted KNN balltree Accuracy = 0.5558
Elapsed time = 0.0074 secs
weighted KNN balltree Accuracy = 0.5297
Elapsed time = 0.0050 secs

balltree unweighted with n =  15
unweighted KNN ball

In [ ]:
#kd_tree weighted and unweighted, with ablility to change n

def kd_tree_knnC(xtr, ytr, xte, yte, n = 15):
  start = time.time()
  model = KNeighborsClassifier(n_neighbors=n, algorithm='kd_tree').fit(xtr, ytr)

  pred = model.predict(xte)
  accuracy = accuracy_score(yte, pred)
  print(f'unweighted KNN kdtree Accuracy = {accuracy:.4}')

  #default_knn_accuracy = default_knn.score(xte,yte)
  #print("Knn Accuracy ", default_knn_accuracy)

  elapsed_time = time.time() - start
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))

def Wkd_tree_knnC(xtr, ytr, xte, yte, n = 15):
  start = time.time()
  model = KNeighborsClassifier(n_neighbors=n, weights = 'distance', algorithm='kd_tree').fit(xtr, ytr)

  pred = model.predict(xte)
  accuracy = accuracy_score(yte, pred)
  print(f'Weighted KNN tdtree Accuracy = {accuracy:.4}')

  #default_knn_accuracy = default_knn.score(xte,yte)
  #print("Knn Accuracy ", default_knn_accuracy)

  elapsed_time = time.time() - start
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))

for n in range(11, 18, 2):
  print("kdtree unweighted with n = ", n)
  kd_tree_knnC(x70_train, y70_train, x30_test, y30_test, n)
  kd_tree_knnC(x80_train, y80_train, x20_test, y20_test, n)
  kd_tree_knnC(x90_train, y90_train, x10_test, y10_test, n)
  print()
  print("kdtree weighted with n = ", n)
  Wkd_tree_knnC(x70_train, y70_train, x30_test, y30_test, n)
  Wkd_tree_knnC(x80_train, y80_train, x20_test, y20_test, n)
  Wkd_tree_knnC(x90_train, y90_train, x10_test, y10_test, n)
  print()

kdtree unweighted with n =  11
unweighted KNN kdtree Accuracy = 0.5762
Elapsed time = 0.0277 secs
unweighted KNN kdtree Accuracy = 0.5806
Elapsed time = 0.0199 secs
unweighted KNN kdtree Accuracy = 0.5693
Elapsed time = 0.0141 secs

kdtree weighted with n =  11
Weighted KNN tdtree Accuracy = 0.5613
Elapsed time = 0.0055 secs
Weighted KNN tdtree Accuracy = 0.5533
Elapsed time = 0.0050 secs
Weighted KNN tdtree Accuracy = 0.5644
Elapsed time = 0.0051 secs

kdtree unweighted with n =  13
unweighted KNN kdtree Accuracy = 0.5579
Elapsed time = 0.0241 secs
unweighted KNN kdtree Accuracy = 0.5682
Elapsed time = 0.0185 secs
unweighted KNN kdtree Accuracy = 0.5594
Elapsed time = 0.0157 secs

kdtree weighted with n =  13
Weighted KNN tdtree Accuracy = 0.5629
Elapsed time = 0.0079 secs
Weighted KNN tdtree Accuracy = 0.5558
Elapsed time = 0.0053 secs
Weighted KNN tdtree Accuracy = 0.5297
Elapsed time = 0.0045 secs

kdtree unweighted with n =  15
unweighted KNN kdtree Accuracy = 0.5629
Elapsed time 

In [ ]:
#worst algorithm

#brute weighted and unweighted

def brute_knnC(xtr, ytr, xte, yte, n = 15):
  start = time.time()
  model = KNeighborsClassifier(n_neighbors=n, algorithm='brute').fit(xtr, ytr)

  pred = model.predict(xte)
  accuracy = accuracy_score(yte, pred)
  print(f'unweighted KNN brute Accuracy = {accuracy:.4}')

  #default_knn_accuracy = default_knn.score(xte,yte)
  #print("Knn Accuracy ", default_knn_accuracy)

  elapsed_time = time.time() - start
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))

def Wbrute_knnC(xtr, ytr, xte, yte, n = 15):
  start = time.time()
  model = KNeighborsClassifier(n_neighbors=n, weights = 'distance', algorithm='brute').fit(xtr, ytr)

  pred = model.predict(xte)
  accuracy = accuracy_score(yte, pred)
  print(f'weighted KNN brute Accuracy = {accuracy:.4}')

  #default_knn_accuracy = default_knn.score(xte,yte)
  #print("Knn Accuracy ", default_knn_accuracy)

  elapsed_time = time.time() - start
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))

for n in range(11, 18, 2):
  print("brute unweighted with n = ", n)
  brute_knnC(x70_train, y70_train, x30_test, y30_test)
  brute_knnC(x80_train, y80_train, x20_test, y20_test)
  brute_knnC(x90_train, y90_train, x10_test, y10_test)

  print("brute weighted with n = ", n)
  Wbrute_knnC(x70_train, y70_train, x30_test, y30_test)
  Wbrute_knnC(x80_train, y80_train, x20_test, y20_test)
  Wbrute_knnC(x90_train, y90_train, x10_test, y10_test)
  print()

brute unweighted with n =  11
unweighted KNN brute Accuracy = 0.5629
Elapsed time = 0.0723 secs
unweighted KNN brute Accuracy = 0.5732
Elapsed time = 0.0493 secs
unweighted KNN brute Accuracy = 0.5644
Elapsed time = 0.0259 secs
brute weighted with n =  11
weighted KNN brute Accuracy = 0.5596
Elapsed time = 0.0236 secs
weighted KNN brute Accuracy = 0.5782
Elapsed time = 0.0185 secs
weighted KNN brute Accuracy = 0.5347
Elapsed time = 0.0138 secs

brute unweighted with n =  13
unweighted KNN brute Accuracy = 0.5629
Elapsed time = 0.0586 secs
unweighted KNN brute Accuracy = 0.5732
Elapsed time = 0.0392 secs
unweighted KNN brute Accuracy = 0.5644
Elapsed time = 0.0295 secs
brute weighted with n =  13
weighted KNN brute Accuracy = 0.5596
Elapsed time = 0.0231 secs
weighted KNN brute Accuracy = 0.5782
Elapsed time = 0.0186 secs
weighted KNN brute Accuracy = 0.5347
Elapsed time = 0.0125 secs

brute unweighted with n =  15
unweighted KNN brute Accuracy = 0.5629
Elapsed time = 0.0607 secs
unweig

In [ ]:
#using this as best for knn classification models 
#weighted knn

def weighted_knn_auto(xtr, ytr, xte, yte, n):
  start = time.time()
  model = KNeighborsClassifier(n_neighbors=n, weights = 'distance').fit(xtr, ytr)
  pred = model.predict(xte)
  accuracy = accuracy_score(yte, pred)
  print('Weighted Knn Accuracy = {:7.4f}'.format(accuracy))

  elapsed_time = time.time() - start
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))
  
n =17
print("weighted knn with n = ", n)
weighted_knn_auto(x70_train, y70_train, x30_test, y30_test, n)
weighted_knn_auto(x80_train, y80_train, x20_test, y20_test, n)
weighted_knn_auto(x90_train, y90_train, x10_test, y10_test, n)
print()
print("balltree weighted with n = ", n)
Wball_tree_knnC(x70_train, y70_train, x30_test, y30_test, n)
Wball_tree_knnC(x80_train, y80_train, x20_test, y20_test, n)
Wball_tree_knnC(x90_train, y90_train, x10_test, y10_test, n)
print()
print("kdtree weighted with n = ", n)
Wkd_tree_knnC(x70_train, y70_train, x30_test, y30_test, n)
Wkd_tree_knnC(x80_train, y80_train, x20_test, y20_test, n)
Wkd_tree_knnC(x90_train, y90_train, x10_test, y10_test, n)
print()

weighted knn with n =  17
Weighted Knn Accuracy =  0.5778
Elapsed time = 0.0094 secs
Weighted Knn Accuracy =  0.5806
Elapsed time = 0.0080 secs
Weighted Knn Accuracy =  0.5446
Elapsed time = 0.0068 secs

balltree weighted with n =  17
weighted KNN balltree Accuracy = 0.5778
Elapsed time = 0.0123 secs
weighted KNN balltree Accuracy = 0.5806
Elapsed time = 0.0103 secs
weighted KNN balltree Accuracy = 0.5446
Elapsed time = 0.0076 secs

kdtree weighted with n =  17
Weighted KNN tdtree Accuracy = 0.5778
Elapsed time = 0.0105 secs
Weighted KNN tdtree Accuracy = 0.5806
Elapsed time = 0.0055 secs
Weighted KNN tdtree Accuracy = 0.5446
Elapsed time = 0.0074 secs



# Best KNN Classification model and parameters

In [ ]:
n = 17
print("weighted knn with n = ", n)
weighted_knn_auto(x80_train, y80_train, x20_test, y20_test, n)
Wball_tree_knnC(x80_train, y80_train, x20_test, y20_test, n)
Wkd_tree_knnC(x80_train, y80_train, x20_test, y20_test, n)

weighted knn with n =  17
Weighted Knn Accuracy =  0.5806
Elapsed time = 0.0078 secs
weighted KNN balltree Accuracy = 0.5806
Elapsed time = 0.0105 secs
Weighted KNN tdtree Accuracy = 0.5806
Elapsed time = 0.0072 secs


# **Knn results and analysis**

Knn-nearest neighbors fast BUT accuracy of this model is inadquate to meet the needs and safety of The World Health Organization since accuracy maxes out on average to .58 using by 80 by 20 dataset, n = 17, using weighted knn, and using either Ball_tree or  kd_tree algorithm to also get the fastest speed vs unweighted knn and/or using brute algorithm. 

# Naive Bayes Classification Model

##**Naive Bayes model**

### Altering binarize and alpha(smoothing)

In [ ]:
#alpha for Multinomial, Complement, and Bernoulli
#changed alpha to .5, vs default 1 
# however, when a =0 (no smoothing),  program changes a = 1.0*e(-10)
a = .5

### Gaussian Naive Bayes

In [ ]:
#Gaussian Naive Bayes default

def default_Gaussian(xtr, ytr, xte, yte):
  start = time.time()
  model = GaussianNB().fit(xtr, ytr)
  pred = model.predict(xte)
  accuracy = accuracy_score(yte, pred)
  print('Gaussian default Accuracy = {:7.4f}'.format(accuracy))

  elapsed_time = time.time() - start
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))
  


default_Gaussian(x70_train, y70_train, x30_test, y30_test)
default_Gaussian(x80_train, y80_train, x20_test, y20_test)
default_Gaussian(x90_train, y90_train, x10_test, y10_test)

Gaussian default Accuracy =  0.6209
Elapsed time = 0.0031 secs
Gaussian default Accuracy =  0.6352
Elapsed time = 0.0019 secs
Gaussian default Accuracy =  0.6238
Elapsed time = 0.0017 secs


### Multinomial Naive Bayes

In [ ]:
#Multinomial Naive Bayes default

def default_multin(xtr, ytr, xte, yte):
  start = time.time()
  model = MultinomialNB().fit(xtr, ytr)
  pred = model.predict(xte)
  accuracy = accuracy_score(yte, pred)
  print('Multinomial Default Accuracy = {:7.4f}'.format(accuracy))

  elapsed_time = time.time() - start
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))
  
default_multin(x70_train, y70_train, x30_test, y30_test)
default_multin(x80_train, y80_train, x20_test, y20_test)
default_multin(x90_train, y90_train, x10_test, y10_test)


Multinomial Default Accuracy =  0.5348
Elapsed time = 0.0049 secs
Multinomial Default Accuracy =  0.5112
Elapsed time = 0.0021 secs
Multinomial Default Accuracy =  0.5149
Elapsed time = 0.0018 secs


In [ ]:
#Multinomial Naive Bayes mods/changes

def alpha_multin(xtr, ytr, xte, yte):
  start = time.time()
  model = MultinomialNB(alpha= a).fit(xtr, ytr)
  pred = model.predict(xte)
  accuracy = accuracy_score(yte, pred)
  print('Alpha = ', a)
  print('Multinomial Accuracy = {:7.4f}'.format(accuracy))

  elapsed_time = time.time() - start
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))
  
alpha_multin(x70_train, y70_train, x30_test, y30_test)
alpha_multin(x80_train, y80_train, x20_test, y20_test)
alpha_multin(x90_train, y90_train, x10_test, y10_test)

Alpha =  0.5
Multinomial Accuracy =  0.5348
Elapsed time = 0.0045 secs
Alpha =  0.5
Multinomial Accuracy =  0.5112
Elapsed time = 0.0025 secs
Alpha =  0.5
Multinomial Accuracy =  0.5149
Elapsed time = 0.0024 secs


### Complement Naive Bayes

In [ ]:
#Complement Naive Bayes default 

def default_compl(xtr, ytr, xte, yte):
  start = time.time()
  model = ComplementNB().fit(xtr, ytr)
  pred = model.predict(xte)
  accuracy = accuracy_score(yte, pred)
  print('Complement Default Accuracy = {:7.4f}'.format(accuracy))

  elapsed_time = time.time() - start
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))
  
default_compl(x70_train, y70_train, x30_test, y30_test)
default_compl(x80_train, y80_train, x20_test, y20_test)
default_compl(x90_train, y90_train, x10_test, y10_test)


Complement Default Accuracy =  0.5298
Elapsed time = 0.0026 secs
Complement Default Accuracy =  0.5062
Elapsed time = 0.0019 secs
Complement Default Accuracy =  0.5099
Elapsed time = 0.0018 secs


In [ ]:
#Complement Naive Bayes mods/changes


def alpha_compl(xtr, ytr, xte, yte):
  start = time.time()
  model = ComplementNB(alpha= a).fit(xtr, ytr)
  pred = model.predict(xte)
  accuracy = accuracy_score(yte, pred)
  print("Alpha = ", a)
  print('Complement Default Accuracy = {:7.4f}'.format(accuracy))

  elapsed_time = time.time() - start
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))
  
alpha_compl(x70_train, y70_train, x30_test, y30_test)
alpha_compl(x80_train, y80_train, x20_test, y20_test)
alpha_compl(x90_train, y90_train, x10_test, y10_test)

Alpha =  0.5
Complement Default Accuracy =  0.5298
Elapsed time = 0.0032 secs
Alpha =  0.5
Complement Default Accuracy =  0.5062
Elapsed time = 0.0024 secs
Alpha =  0.5
Complement Default Accuracy =  0.5099
Elapsed time = 0.0024 secs


### Bernoulli Naive Bayes

In [ ]:
#Bernoulli Naive Bayes default

def default_bern(xtr, ytr, xte, yte):
  start = time.time()
  model = BernoulliNB().fit(xtr, ytr)
  pred = model.predict(xte)
  accuracy = accuracy_score(yte, pred)
  print('Bernouli Default Accuracy = {:7.4f}'.format(accuracy))

  elapsed_time = time.time() - start
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))
  
default_bern(x70_train, y70_train, x30_test, y30_test)
default_bern(x80_train, y80_train, x20_test, y20_test)
default_bern(x90_train, y90_train, x10_test, y10_test)


Bernouli Default Accuracy =  0.5960
Elapsed time = 0.0039 secs
Bernouli Default Accuracy =  0.5931
Elapsed time = 0.0045 secs
Bernouli Default Accuracy =  0.5743
Elapsed time = 0.0033 secs


In [ ]:
#Bernoulli Naive Bayes mods/changes bin

def mod_bern(xtr, ytr, xte, yte, a = 1, b= .5):
  start = time.time()
  model = BernoulliNB(alpha = a, binarize= b).fit(xtr, ytr)
  pred = model.predict(xte)
  accuracy = accuracy_score(yte, pred)
  print("alpha = ", a)
  print("binarize = ", b)
  print('Bernouli mod Accuracy = {:7.4f}'.format(accuracy))

  elapsed_time = time.time() - start
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))

# tried np.mean(x_train) and .5 for binarize for Bernoulli
#bin = .5

#bin7m = np.mean(x70_train)
#bin8m = np.mean(x80_train)
#bin9m = np.mean(x90_train)
mod_bern(x70_train, y70_train, x30_test, y30_test)
mod_bern(x80_train, y80_train, x20_test, y20_test)
mod_bern(x90_train, y90_train, x10_test, y10_test)


alpha =  1
binarize =  0.5
Bernouli mod Accuracy =  0.5960
Elapsed time = 0.0045 secs
alpha =  1
binarize =  0.5
Bernouli mod Accuracy =  0.5931
Elapsed time = 0.0071 secs
alpha =  1
binarize =  0.5
Bernouli mod Accuracy =  0.5743
Elapsed time = 0.0031 secs


# Best Naive Bayes model and parameters

In [ ]:
def bestNB_Gaussian(xtr, ytr, xte, yte):
  start = time.time()
  model = GaussianNB().fit(xtr, ytr)
  pred = model.predict(xte)
  accuracy = accuracy_score(yte, pred)
  print('Gaussian default Accuracy = {:7.4f}'.format(accuracy))

  elapsed_time = time.time() - start
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))

# Accuracy and speed results of Naive Bayes 
**BEST**

**Gaussian Naive Bayes** with default parameters was the best model when paired with a dataset split into 80 % train  by 20% test with an accuracy of **.63** and speed .0035 sec. 

**Burn** was second place when paired with 70%train by 30%test for best accuracy which was **.59**, while Best parameters to help increase speed(not accuracy) was setting  alpha = .5 and binarize = mean(x_train) to gain an average speed of .0046 sec.

**Multinomial Naive Bayes** was third place when paired with 70%train by 30%test for best accuracy of .53 . Best parameters for it was a = .5 to gain an average speed of .0042

**WORST**

**Complement Naive Bayes** was last place when paired with 70%train by 30%test for best accuracy of .52 . Best parameters for it was not significant changes found when change alpha for this.


#**Decision Tree model**

### Using Gini

In [ ]:
# Decison Tree Default with Gini

#This is where it hits me that I could have saved time and code space
#by defining functions 

def default_G_DTC(xtr, ytr, xte, yte):
  start = time.time()
  model = DecisionTreeClassifier().fit(xtr, ytr)
  pred = model.predict(xte)
  accuracy = accuracy_score(yte, pred)
  print('Accuracy = {:7.4f}'.format(accuracy))

  elapsed_time = time.time() - start
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))
  
  #cm = confusion_matrix(yte, pred)
  #print(cm)


default_G_DTC(x70_train, y70_train, x30_test, y30_test)
default_G_DTC(x80_train, y80_train, x20_test, y20_test)
default_G_DTC(x90_train, y90_train, x10_test, y10_test)

Accuracy =  0.6043
Elapsed time = 0.0191 secs
Accuracy =  0.6055
Elapsed time = 0.0209 secs
Accuracy =  0.5594
Elapsed time = 0.0238 secs


look for best depth

In [ ]:
# Decison Tree with Gini looking for best depth 
#

def best_G_DTC_depth(xtr, ytr, xte, yte):
  best_a = 0
  best_d = 0
  speed = 0

  #find max_depth that gets best results best split
  for d in range(1,17):
    start = time.time()
    model = DecisionTreeClassifier(max_depth = d).fit(xtr, ytr)
    pred = model.predict(xte)
    accuracy = accuracy_score(yte, pred)
    #print('Max_depth = {};, accuracy = {:7.4f}'.format(d ,accuracy))

    elapsed_time = time.time() - start
    #print('Elapsed time = {:.4f} secs'.format(elapsed_time))

    if(best_a < accuracy):
      best_a = accuracy
      best_d = d
      speed = elapsed_time
  print('Max_depth = {};, B_accuracy = {:7.4f}'.format(best_d , best_a))
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))

def rand_G_DTC_depth(xtr, ytr, xte, yte):
  best_a = 0
  best_d = 0
  speed = 0

  #find max_depth that gets best results rand split
  for d in range(1,16):
    start = time.time()
    model = DecisionTreeClassifier(splitter = 'random', max_depth = d).fit(xtr, ytr)
    pred = model.predict(xte)
    accuracy = accuracy_score(yte, pred)
    #print('Max_depth = {};, accuracy = {:7.4f}'.format(d ,accuracy))

    elapsed_time = time.time() - start
    #print('Elapsed time = {:.4f} secs'.format(elapsed_time))

    if(best_a < accuracy):
      best_a = accuracy
      best_d = d
      speed = elapsed_time
  print('Max_depth = {};, R_accuracy = {:7.4f}'.format(best_d , best_a))
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))

print('Best splitter, find max depth')
best_G_DTC_depth(x70_train, y70_train, x30_test, y30_test)
best_G_DTC_depth(x80_train, y80_train, x20_test, y20_test)
best_G_DTC_depth(x90_train, y90_train, x10_test, y10_test)

print()
print('Random splitter, find max depth')
rand_G_DTC_depth(x70_train, y70_train, x30_test, y30_test)
rand_G_DTC_depth(x80_train, y80_train, x20_test, y20_test)
rand_G_DTC_depth(x90_train, y90_train, x10_test, y10_test)

Best splitter, find max depth
Max_depth = 7;, B_accuracy =  0.6507
Elapsed time = 0.0180 secs
Max_depth = 5;, B_accuracy =  0.6452
Elapsed time = 0.0180 secs
Max_depth = 6;, B_accuracy =  0.6485
Elapsed time = 0.0209 secs

Random splitter, find max depth
Max_depth = 5;, R_accuracy =  0.6440
Elapsed time = 0.0032 secs
Max_depth = 11;, R_accuracy =  0.6551
Elapsed time = 0.0031 secs
Max_depth = 6;, R_accuracy =  0.6485
Elapsed time = 0.0034 secs


Use depth, random split, and random_state

In [ ]:
# Decison Tree with Gini looking for best depth 
# with random split and random state
#random state, tried 0 and 42 (popular and most common)

 #find max_depth that gets best results with random split and random st
def final_G_DTCm1(xtr, ytr, xte, yte, ran_state = 0):
  best_a = 0
  best_d = 0
  speed = 0

  for d in range(1,20):
    start = time.time()
    model = DecisionTreeClassifier(max_depth = d, random_state = ran_state).fit(xtr, ytr)
    pred = model.predict(xte)
    accuracy = accuracy_score(yte, pred)
    #print('Max_depth = {};, accuracy = {:7.4f}'.format(d ,accuracy))

    elapsed_time = time.time() - start
    #print('Elapsed time = {:.4f} secs'.format(elapsed_time))

    if(best_a < accuracy):
      best_a = accuracy
      best_d = d
      speed = elapsed_time
  print('Max_depth = {};, accuracy = {:7.4f}'.format(best_d , best_a))
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))


#splitter
def final_G_DTCm2(xtr, ytr, xte, yte, ran_state = 0):
  best_a = 0
  best_d = 0
  speed = 0

  for d in range(1,20):
    start = time.time()
    model = DecisionTreeClassifier(splitter = 'random', max_depth = d, random_state = ran_state).fit(xtr, ytr)
    pred = model.predict(xte)
    accuracy = accuracy_score(yte, pred)
    #print('Max_depth = {};, accuracy = {:7.4f}'.format(d ,accuracy))

    elapsed_time = time.time() - start
    #print('Elapsed time = {:.4f} secs'.format(elapsed_time))

    if(best_a < accuracy):
      best_a = accuracy
      best_d = d
      speed = elapsed_time
  print('Max_depth = {};, accuracy = {:7.4f}'.format(best_d , best_a))
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))

r_state = 42
final_G_DTCm1(x70_train, y70_train, x30_test, y30_test, r_state)
final_G_DTCm1(x80_train, y80_train, x20_test, y20_test, r_state)
final_G_DTCm1(x90_train, y90_train, x10_test, y10_test, r_state)
print()
final_G_DTCm2(x70_train, y70_train, x30_test, y30_test, r_state)
final_G_DTCm2(x80_train, y80_train, x20_test, y20_test, r_state)
final_G_DTCm2(x90_train, y90_train, x10_test, y10_test, r_state)

Max_depth = 7;, accuracy =  0.6507
Elapsed time = 0.0165 secs
Max_depth = 9;, accuracy =  0.6501
Elapsed time = 0.0229 secs
Max_depth = 6;, accuracy =  0.6436
Elapsed time = 0.0236 secs

Max_depth = 7;, accuracy =  0.6573
Elapsed time = 0.0032 secs
Max_depth = 4;, accuracy =  0.6551
Elapsed time = 0.0042 secs
Max_depth = 4;, accuracy =  0.6634
Elapsed time = 0.0034 secs


In [ ]:
# Decison Tree with Gini looking for best depth 
#add max_features
# with random split and random state
#random state, tried 0 and 42 (popular and most common)

 #find max_depth that gets best results with random split and random st
def final_G_DTCm3(xtr, ytr, xte, yte, mf):
  best_a = 0
  best_d = 0
  speed = 0
  
  for d in range(1,18):
    start = time.time()
    model = DecisionTreeClassifier(max_depth = d, random_state = 42, max_features=mf).fit(xtr, ytr)
    pred = model.predict(xte)
    accuracy = accuracy_score(yte, pred)
    #print('Max_depth = {};, accuracy = {:7.4f}'.format(d ,accuracy))

    elapsed_time = time.time() - start
    #print('Elapsed time = {:.4f} secs'.format(elapsed_time))

    if(best_a < accuracy):
      best_a = accuracy
      best_d = d
      speed = elapsed_time
  print('Max_depth = {};, accuracy = {:7.4f}'.format(best_d , best_a))
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))


#splitter
def final_G_DTCm4(xtr, ytr, xte, yte, mf):
  best_a = 0
  best_d = 0
  speed = 0

  for d in range(1,18):
    start = time.time()
    model = DecisionTreeClassifier(splitter = 'random', max_depth = d, random_state = 42, max_features = mf).fit(xtr, ytr)
    pred = model.predict(xte)
    accuracy = accuracy_score(yte, pred)
    #print('Max_depth = {};, accuracy = {:7.4f}'.format(d ,accuracy))

    elapsed_time = time.time() - start
    #print('Elapsed time = {:.4f} secs'.format(elapsed_time))

    if(best_a < accuracy):
      best_a = accuracy
      best_d = d
      speed = elapsed_time
  print('Max_depth = {};, accuracy = {:7.4f}'.format(best_d , best_a))
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))

for mff in range(1,10):
  print("Max_f = ", mff)
  print()
  print("best splitter")
  final_G_DTCm3(x70_train, y70_train, x30_test, y30_test, mff)
  final_G_DTCm3(x80_train, y80_train, x20_test, y20_test, mff)
  final_G_DTCm3(x90_train, y90_train, x10_test, y10_test, mff)
  print()


for mff in range(1,10):
  print("Max_f = ", mff)
  print()
  print("random splitter")
  final_G_DTCm4(x70_train, y70_train, x30_test, y30_test, mff)
  final_G_DTCm4(x80_train, y80_train, x20_test, y20_test, mff)
  final_G_DTCm4(x90_train, y90_train, x10_test, y10_test, mff)
  print()
  

Max_f =  1

best splitter
Max_depth = 7;, accuracy =  0.6275
Elapsed time = 0.0034 secs
Max_depth = 8;, accuracy =  0.6551
Elapsed time = 0.0036 secs
Max_depth = 5;, accuracy =  0.6238
Elapsed time = 0.0062 secs

Max_f =  2

best splitter
Max_depth = 6;, accuracy =  0.6358
Elapsed time = 0.0051 secs
Max_depth = 9;, accuracy =  0.6501
Elapsed time = 0.0067 secs
Max_depth = 13;, accuracy =  0.6931
Elapsed time = 0.0064 secs

Max_f =  3

best splitter
Max_depth = 14;, accuracy =  0.6623
Elapsed time = 0.0065 secs
Max_depth = 10;, accuracy =  0.6824
Elapsed time = 0.0084 secs
Max_depth = 7;, accuracy =  0.6535
Elapsed time = 0.0124 secs

Max_f =  4

best splitter
Max_depth = 7;, accuracy =  0.6556
Elapsed time = 0.0114 secs
Max_depth = 9;, accuracy =  0.6526
Elapsed time = 0.0098 secs
Max_depth = 4;, accuracy =  0.6287
Elapsed time = 0.0136 secs

Max_f =  5

best splitter
Max_depth = 7;, accuracy =  0.6722
Elapsed time = 0.0130 secs
Max_depth = 5;, accuracy =  0.6427
Elapsed time = 0.0119 

Best results with Gini is with paramters best splitter, random state 42, and setting the max_feature to 3, and max_depth at 10

### Using Entropy

In [ ]:
# Decison Tree Default with entropy

def default_E_DTC(xtr, ytr, xte, yte):
  #m1_acc = 0.0
  
  #since results are randomized get a mean/average of 30 default
  for i in range (1):  
    start = time.time()
    
    model = DecisionTreeClassifier(criterion='entropy').fit(xtr, ytr) 
    pred = model.predict(xte)
   # m1_acc += accuracy_score(yte, pred)

    accuracy = accuracy_score(yte, pred)
    print('Accuracy = {:7.4f}'.format(accuracy))
 # print('Average accuracy', m1_acc/30)
    elapsed_time = time.time() - start
    print('Elapsed time = {:.4f} secs'.format(elapsed_time))  
  

default_E_DTC(x70_train, y70_train, x30_test, y30_test)
default_E_DTC(x80_train, y80_train, x20_test, y20_test)
default_E_DTC(x90_train, y90_train, x10_test, y10_test)


Accuracy =  0.6076
Elapsed time = 0.0321 secs
Accuracy =  0.6526
Elapsed time = 0.0471 secs
Accuracy =  0.5693
Elapsed time = 0.0484 secs


In [ ]:
# Decison Tree with entropy, look for best depth

def E_DTC_depth(xtr, ytr, xte, yte):
  best_a = 0
  best_d = 0
  speed = 0
  
  for d in range(1,25):
    start = time.time()
    model = DecisionTreeClassifier(criterion='entropy', max_depth = d).fit(xtr, ytr)
    pred = model.predict(xte)
    accuracy = accuracy_score(yte, pred)
    #print('Max_depth = {};, accuracy = {:7.4f}'.format(d ,accuracy))

    elapsed_time = time.time() - start
    #print('Elapsed time = {:.4f} secs'.format(elapsed_time))

    if(best_a < accuracy):
      best_a = accuracy
      best_d = d
      speed = elapsed_time
  print('Max_depth = {};, accuracy = {:7.4f}'.format(best_d , best_a))
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))

E_DTC_depth(x70_train, y70_train, x30_test, y30_test)
E_DTC_depth(x80_train, y80_train, x20_test, y20_test)
E_DTC_depth(x90_train, y90_train, x10_test, y10_test)

Max_depth = 13;, accuracy =  0.6407
Elapsed time = 0.0310 secs
Max_depth = 14;, accuracy =  0.6675
Elapsed time = 0.0418 secs
Max_depth = 6;, accuracy =  0.6436
Elapsed time = 0.0448 secs


The process to find best model for decision tree

In [ ]:
# Decison Tree with entropy, look for best depth 
#try splitter and random state (0 or 42) parameters
#dont use random split, lowers accuracy still

def E_DTC_m1(xtr, ytr, xte, yte, mf):
  best_a = 0
  best_d = 0
  speed = 0
  
  for d in range(4,20):
    start = time.time()
    model = DecisionTreeClassifier(criterion='entropy',  max_depth = d, random_state = 42, max_features=mf).fit(xtr, ytr)
    pred = model.predict(xte)
    accuracy = accuracy_score(yte, pred)
    #print('Max_depth = {};, accuracy = {:7.4f}'.format(d ,accuracy))

    elapsed_time = time.time() - start
    #print('Elapsed time = {:.4f} secs'.format(elapsed_time))

    if(best_a < accuracy):
      best_a = accuracy
      best_d = d
      speed = elapsed_time
  print('Max_depth = {};, accuracy = {:7.4f}'.format(best_d , best_a))
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))

def E_DTC_m2(xtr, ytr, xte, yte, mf):
  best_a = 0
  best_d = 0
  speed = 0

  for d in range(4,20):
    start = time.time()
    model = DecisionTreeClassifier(criterion='entropy', splitter = 'random' ,  max_depth = d, random_state = 42, max_features=mf).fit(xtr, ytr)
    pred = model.predict(xte)
    accuracy = accuracy_score(yte, pred)
    #print('Max_depth = {};, accuracy = {:7.4f}'.format(d ,accuracy))

    elapsed_time = time.time() - start
    #print('Elapsed time = {:.4f} secs'.format(elapsed_time))

    if(best_a < accuracy):
      best_a = accuracy
      best_d = d
      speed = elapsed_time
  print('Max_depth = {};, accuracy = {:7.4f}'.format(best_d , best_a))
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))



for maxf in range(1,10):
  print("Max_f = ", maxf)
  print()
  print("best splitter")
  E_DTC_m1(x70_train, y70_train, x30_test, y30_test, maxf)
  E_DTC_m1(x80_train, y80_train, x20_test, y20_test, maxf)
  E_DTC_m1(x90_train, y90_train, x10_test, y10_test, maxf)
  print()


for maxf in range(1,10):
  print("Max_f = ", maxf)
  print()
  print("rand split")
  E_DTC_m2(x70_train, y70_train, x30_test, y30_test, maxf)
  E_DTC_m2(x80_train, y80_train, x20_test, y20_test, maxf)
  E_DTC_m2(x90_train, y90_train, x10_test, y10_test, maxf)
  print()

Max_f =  1

best splitter
Max_depth = 7;, accuracy =  0.6358
Elapsed time = 0.0057 secs
Max_depth = 8;, accuracy =  0.6526
Elapsed time = 0.0065 secs
Max_depth = 14;, accuracy =  0.6337
Elapsed time = 0.0081 secs

Max_f =  2

best splitter
Max_depth = 8;, accuracy =  0.6474
Elapsed time = 0.0100 secs
Max_depth = 11;, accuracy =  0.6625
Elapsed time = 0.0137 secs
Max_depth = 11;, accuracy =  0.6782
Elapsed time = 0.0131 secs

Max_f =  3

best splitter
Max_depth = 13;, accuracy =  0.6523
Elapsed time = 0.0143 secs
Max_depth = 8;, accuracy =  0.6476
Elapsed time = 0.0172 secs
Max_depth = 6;, accuracy =  0.6584
Elapsed time = 0.0202 secs

Max_f =  4

best splitter
Max_depth = 7;, accuracy =  0.6507
Elapsed time = 0.0141 secs
Max_depth = 19;, accuracy =  0.6700
Elapsed time = 0.0200 secs
Max_depth = 11;, accuracy =  0.6683
Elapsed time = 0.0229 secs

Max_f =  5

best splitter
Max_depth = 10;, accuracy =  0.6573
Elapsed time = 0.0179 secs
Max_depth = 9;, accuracy =  0.6650
Elapsed time = 0.0

In [ ]:
def best_dtc(xtr, ytr, xte, yte, d = 9):
  start = time.time()
  model = DecisionTreeClassifier(criterion='entropy',  max_depth = d, random_state = 42, max_features=8).fit(xtr, ytr)
  pred = model.predict(xte)
  accuracy = accuracy_score(yte, pred)
  print(f'Accuracy = {accuracy:.4}')

  elapsed_time = time.time() - start
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))

best_dtc(x70_train, y70_train, x30_test, y30_test)
best_dtc(x80_train, y80_train, x20_test, y20_test)
best_dtc(x90_train, y90_train, x10_test, y10_test)

Accuracy = 0.6507
Elapsed time = 0.0284 secs
Accuracy = 0.6898
Elapsed time = 0.0265 secs
Accuracy = 0.5644
Elapsed time = 0.0292 secs


# Accuracy and speed results of Decision Tree

**Best**

The best CONSISTENT decision tree results were accuracy = .6898 with speed = .0254 secs at max_depth = 9, random_state = 42, max_features = 8  using entropy over gini, and with a training set of 80% training and 20%test.

**Worst**
 adding/injecting randomness will hurt accuracy and give inconsistent depth/max depth in which the best accuracy can be found IF not paired with applied the random_state parameters.  Experienced all the decision tree tests had a noticable decrease in accuracy by overfitting when dataset was split in 90% train and 10% test  

# **Best models for classifers with their paramaters** 

In [ ]:
#knn
print("Best knn model")
print("80% train by 20% test")
weighted_knn_auto(x80_train, y80_train, x20_test, y20_test, 15)

#naive bayes
print()
print("Best Naive Baye model")
print("80% train by 20% test")
bestNB_Gaussian(x80_train, y80_train, x20_test, y20_test)


#Decision Trees
print()
print("Best Decision tree model")
print("80% train by 20% test")
best_dtc(x80_train, y80_train, x20_test, y20_test)


Best knn model
80% train by 20% test
Weighted Knn Accuracy =  0.5782
Elapsed time = 0.0112 secs

Best Naive Baye model
80% train by 20% test
Gaussian default Accuracy =  0.6352
Elapsed time = 0.0031 secs

Best Decision tree model
80% train by 20% test
Accuracy = 0.6898
Elapsed time = 0.0270 secs


Best of the best is decision tree with accuracy of .6898, depth 9, splitter = 'best',  max_feature = 8, random_state= 42, and  and trained by the 80% by 20% dataset

 Accuracy = 0.6898
Elapsed time = 0.0257 secs

# Test out Regressors based on mean squared error

1. knn regressor default vs knn regressor mod_paramters

2. Naive Bayes default vs Naive Bayes mod_paramters

3. Decision Tree regressor default vs decision tree regressor mod_paramters 

#**Knn Regressor Model**

In [ ]:
#Default knn Regressor vs default weight knn regressor
## k-nn regressor version
# n_jobs=-1 use all processors
#n = 5 be fault

def default_knnReg(xtr, ytr, xte, yte):
  start = time.time()
  model = KNeighborsRegressor(n_jobs=-1).fit(xtr, ytr)
  pred = model.predict(xte)
  print('Mean squared error = {:5.2f}'.format(mean_squared_error(pred, yte)))
  elapsed_time = time.time() - start
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))

def defaultW_knnReg(xtr, ytr, xte, yte):
  start = time.time() 
  model = KNeighborsRegressor(weights = 'distance', n_jobs=-1).fit(xtr, ytr)
  pred = model.predict(xte)
  print('Mean squared error = {:5.2f}'.format(mean_squared_error(pred, yte)))
  elapsed_time = time.time() - start
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))

print('unweighted knn regressor default')
defaultW_knnReg(x70_train, y70_train, x30_test, y30_test)
defaultW_knnReg(x80_train, y80_train, x20_test, y20_test)
defaultW_knnReg(x90_train, y90_train, x10_test, y10_test)
print()

print('weighted knn regressor default')
defaultW_knnReg(x70_train, y70_train, x30_test, y30_test)
defaultW_knnReg(x80_train, y80_train, x20_test, y20_test)
defaultW_knnReg(x90_train, y90_train, x10_test, y10_test)

unweighted knn regressor default


NameError: ignored

In [ ]:
#change n and algoriths

def auto_knnReg(xtr, ytr, xte, yte, n):
  start = time.time()
  model = KNeighborsRegressor(n_neighbors= n, n_jobs=-1).fit(xtr, ytr)
  pred = model.predict(xte)
  print('Mean squared error = {:5.2f}'.format(mean_squared_error(pred, yte)))

  elapsed_time = time.time() - start
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))


def bt_knnReg(xtr, ytr, xte, yte, n):
  start = time.time()  
  model = KNeighborsRegressor(n_neighbors = n, algorithm = 'ball_tree' ,n_jobs=-1).fit(xtr, ytr)
  pred = model.predict(xte)
  print('Mean squared error = {:5.2f}'.format(mean_squared_error(pred, yte)))

  elapsed_time = time.time() - start
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))


def kd_knnReg(xtr, ytr, xte, yte, n):
  start = time.time()  
  model = KNeighborsRegressor(n_neighbors = n, algorithm = 'kd_tree' ,n_jobs=-1).fit(xtr, ytr)
  pred = model.predict(xte)
  print('Mean squared error = {:5.2f}'.format(mean_squared_error(pred, yte)))

  elapsed_time = time.time() - start
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))


def brute_knnReg(xtr, ytr, xte, yte, n):
  start = time.time()  
  model = KNeighborsRegressor(n_neighbors = n, algorithm = 'brute' ,n_jobs=-1).fit(xtr, ytr)
  pred = model.predict(xte)
  print('Mean squared error = {:5.2f}'.format(mean_squared_error(pred, yte)))
  
  elapsed_time = time.time() - start
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))

#check n=15 from previous results with classifers to see if n =15 means 
#something for knn models

# n= 15 for all dataset models to get .25 (lowest MSE) 
# n = 9, dataset of 90by 10 hits and maintains .25 first


for n in range(7,12,2):
  print('n = ', n)
  print("auto algorithm")
  auto_knnReg(x70_train, y70_train, x30_test, y30_test, n)
  auto_knnReg(x80_train, y80_train, x20_test, y20_test, n)
  auto_knnReg(x90_train, y90_train, x10_test, y10_test, n)

  print()
  print("ball_tree algorithm")
  bt_knnReg(x70_train, y70_train, x30_test, y30_test, n)
  bt_knnReg(x80_train, y80_train, x20_test, y20_test, n)
  bt_knnReg(x90_train, y90_train, x10_test, y10_test, n)

  print()
  print("kd_tree algorithm")
  kd_knnReg(x70_train, y70_train, x30_test, y30_test, n)
  kd_knnReg(x80_train, y80_train, x20_test, y20_test, n)
  kd_knnReg(x90_train, y90_train, x10_test, y10_test, n)

  print()
#print("brute algorithm")
#brute_knnReg(x70_train, y70_train, x30_test, y30_test, n4)
#brute_knnReg(x80_train, y80_train, x20_test, y20_test, n4)
#brute_knnReg(x90_train, y90_train, x10_test, y10_test, n4)


n =  7
auto algorithm


NameError: ignored

In [ ]:
#weighted resulted for all

def autoW_knnReg(xtr, ytr, xte, yte, n):
  start = time.time()  
  model = KNeighborsRegressor(n_neighbors= n, weights = 'distance', n_jobs=-1).fit(xtr, ytr)
  pred = model.predict(xte)
  print('Mean squared error = {:5.2f}'.format(mean_squared_error(pred, yte)))
  
  elapsed_time = time.time() - start
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))


def btW_knnReg(xtr, ytr, xte, yte, n):
  start = time.time()  
  model = KNeighborsRegressor(n_neighbors = n, weights = 'distance', algorithm = 'ball_tree' ,n_jobs=-1).fit(xtr, ytr)
  pred = model.predict(xte)
  print('Mean squared error = {:5.2f}'.format(mean_squared_error(pred, yte)))
  
  elapsed_time = time.time() - start
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))


def kdW_knnReg(xtr, ytr, xte, yte, n):
  start = time.time()  
  model = KNeighborsRegressor(n_neighbors = n, weights = 'distance', algorithm = 'kd_tree' ,n_jobs=-1).fit(xtr, ytr)
  pred = model.predict(xte)
  print('Mean squared error = {:5.2f}'.format(mean_squared_error(pred, yte)))
  
  
  elapsed_time = time.time() - start
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))


def bruteW_knnReg(xtr, ytr, xte, yte, n):
  start = time.time()  
  model = KNeighborsRegressor(n_neighbors = n, weights = 'distance', algorithm = 'brute' ,n_jobs=-1).fit(xtr, ytr)
  pred = model.predict(xte)
  print('Mean squared error = {:5.2f}'.format(mean_squared_error(pred, yte)))
  
  
  elapsed_time = time.time() - start
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))


#check n from previous results with classifers
# n is increasing in order for models to hit the lowest of .25 unlike
#unweighted, thus weighted affects MSE and requires n to increase to 
#get the lowest MSE possible that unweighted got at n = 9, with 90by10

#n= 11 hits the .25 with 90by10 first 
#n = 23 is when all datasets get .25
for n in range(7,24,2):
  print('n = ', n)
  print("Weighted ball_tree algorithm")
  btW_knnReg(x70_train, y70_train, x30_test, y30_test, n)
  btW_knnReg(x80_train, y80_train, x20_test, y20_test, n)
  btW_knnReg(x90_train, y90_train, x10_test, y10_test, n)
  print()
  print("Weighted kd_tree algorithm")
  kdW_knnReg(x70_train, y70_train, x30_test, y30_test, n)
  kdW_knnReg(x80_train, y80_train, x20_test, y20_test, n)
  kdW_knnReg(x90_train, y90_train, x10_test, y10_test, n)
  print()
  print("WEIGHTED auto algorithm")
  autoW_knnReg(x70_train, y70_train, x30_test, y30_test, n)
  autoW_knnReg(x80_train, y80_train, x20_test, y20_test, n)
  autoW_knnReg(x90_train, y90_train, x10_test, y10_test, n) 
  print()

#print()
#print("weighted brute algorithm")
#bruteW_knnReg(x70_train, y70_train, x30_test, y30_test, n)
#bruteW_knnReg(x80_train, y80_train, x20_test, y20_test, n)
#bruteW_knnReg(x90_train, y90_train, x10_test, y10_test, n)

it seems as long as it is same n being used on any alorithm, we get the same MSE and the parameter that effects MSE is when we apply weights/distance to the algorithms. We get same MSE  through out and n = 15 was good for classifer knn however, by using a dataset split into 90by10, the regressor's best is n = 9 with uniform weights. When weighted predictions are applied in the regressor, n needs to increase to get the same lowest mse.  

for speed, on average the 3 alogrithms to look at are auto, kd_tree, and ball_tree since they average about the same between .1060 to .1070. However, kd_tree seems to have a bit more consistency in speed.


Turns out the best result for MSE with KNN regressor is when n = 9, unweighted predictions, dataset split of 90%by10%. algorithm had no effect. 

# MSE and speed results of KNN Regressor Model

Best KNN regressor model is n = 9 with unweighted predictions with lowest MSE of .25 and to get the best speed its a tie between auto, ball_tree, and kd_tree averaging in .1065 but kd_tree has more consistent speed results. 

#**Decision Tree Regressor model**

Default 

In [ ]:
#default decision tree regressor best or random 

def default_DTR(xtr, ytr, xte, yte):
 
  start = time.time()
  model = DecisionTreeRegressor(random_state=42).fit(xtr, ytr) #to get consistent results
  pred= model.predict(xte)

  print('Mean squared error = {:5.2f}'.format(mean_squared_error(pred,yte)))
  elapsed_time = time.time() - start
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))


def defaultR_DTR(xtr, ytr, xte, yte):
   
  start = time.time()
  model = DecisionTreeRegressor(splitter = 'random', random_state=42).fit(xtr, ytr)
  pred= model.predict(xte)
    
  print('Mean squared error = {:5.2f}'.format(mean_squared_error(pred,yte)))
  elapsed_time = time.time() - start
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))
  

print("best split")
default_DTR(x70_train, y70_train, x30_test, y30_test)
default_DTR(x80_train, y80_train, x20_test, y20_test)
default_DTR(x90_train, y90_train, x10_test, y10_test)

print("random")
defaultR_DTR(x70_train, y70_train, x30_test, y30_test)
defaultR_DTR(x80_train, y80_train, x20_test, y20_test)
defaultR_DTR(x90_train, y90_train, x10_test, y10_test)


best split
Mean squared error =  0.37
Elapsed time = 0.0153 secs
Mean squared error =  0.38
Elapsed time = 0.0192 secs
Mean squared error =  0.44
Elapsed time = 0.0209 secs
random
Mean squared error =  0.38
Elapsed time = 0.0039 secs
Mean squared error =  0.41
Elapsed time = 0.0039 secs
Mean squared error =  0.44
Elapsed time = 0.0044 secs


Bagging introduced

In [ ]:
#TREE WITH BAG best results were from sklearn library with random_state =42
#messing with max featueres isnt best for this because best result is when 
#all features are used 

def bagr_DTR(xtr, ytr, xte, yte, rand_state = 0):

    start = time.time()
    model = BaggingRegressor(random_state=rand_state).fit(xtr, ytr)
    pred = model.predict(xte)
    print('Mean squared error = {:5.2f}'.format(mean_squared_error(pred,yte)))
    elapsed_time = time.time() - start
    print('Elapsed time = {:.4f} secs'.format(elapsed_time))
  


print("bagging regressor, random state = 0")
print("70/30")
print()
bagr_DTR(x70_train, y70_train, x30_test, y30_test)
print("80/20")
bagr_DTR(x80_train, y80_train, x20_test, y20_test)
print("90/10")
bagr_DTR(x90_train, y90_train, x10_test, y10_test)

print()
r_st = 42
print("bagging regressor, random state = ", r_st)
print()
print("70/30")
bagr_DTR(x70_train, y70_train, x30_test, y30_test, r_st)
print("80/20")
bagr_DTR(x80_train, y80_train, x20_test, y20_test, r_st)#best result
print("90/10")
bagr_DTR(x90_train, y90_train, x10_test, y10_test, r_st)

bagging regressor, random state = 0
70/30

Mean squared error =  0.22
Elapsed time = 0.1012 secs
80/20
Mean squared error =  0.22
Elapsed time = 0.1065 secs
90/10
Mean squared error =  0.22
Elapsed time = 0.1456 secs

bagging regressor, random state =  42

70/30
Mean squared error =  0.23
Elapsed time = 0.1038 secs
80/20
Mean squared error =  0.21
Elapsed time = 0.1175 secs
90/10
Mean squared error =  0.25
Elapsed time = 0.1419 secs


In [ ]:
#best dtr bagging regressor from library since by default, the base estimator 
#used is a decision tree regressor

def best_dtrbag(xtr, ytr, xte, yte, rand_state = 0):
  
  start = time.time()
  model = BaggingRegressor(random_state=rand_state).fit(xtr, ytr)
  pred = model.predict(xte)
  print('Mean squared error = {:5.2f}'.format(mean_squared_error(pred,yte)))
  elapsed_time = time.time() - start
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))
 
rand_state = 42
print("best results using bagging regressor library")
best_dtrbag(x80_train, y80_train, x20_test, y20_test, rand_state)

best results using bagging regressor library
Mean squared error =  0.21
Elapsed time = 0.1125 secs


Get best depth, alter random state, max_features, and criterion of split, 
and test between best vs random splitter

In [ ]:
#criterion = 'mse', friedman_mse, no difference
#max_features=None (column based, can go up to 9, this is related to attributes)
#kept changing other paramters values and max_depth was still 4

def bes_DTR(xtr, ytr, xte, yte, rand_state = 0):
  best_mse = 0
  best_d = 0
  speed = 0

  for bestMF in range (1,9):  
    print("bestmf: ",bestMF)
    for depth in range (1,8):  
      start = time.time()
      model = DecisionTreeRegressor(max_depth = depth, random_state = rand_state, max_features=bestMF).fit(xtr, ytr)
      pred = model.predict(xte)
      mse = mean_squared_error(pred,yte)

    
      #print('Mean squared error = {:5.2f}'.format(mean_squared_error(pred,yte)))
      elapsed_time = time.time() - start
      #print('Elapsed time = {:.4f} secs'.format(elapsed_time))
      if(best_mse < mse):
        best_mse = mse
        best_d = depth
        speed = elapsed_time

    print('Max_depth = {};, R_mse = {:7.4f}'.format(best_d , best_mse))
    print('Elapsed time = {:.4f} secs'.format(speed))
    print()


def rand_DTR(xtr, ytr, xte, yte, rand_state = 0):
  best_mse = 0
  best_d = 0
  speed = 0

  for bestMF in range (1,xtr.shape[1]):
    print("bestmf: ",bestMF)
    for depth in range (1,8):   
      start = time.time()
      model = DecisionTreeRegressor(splitter='random', max_depth = depth, random_state = rand_state, max_features=bestMF).fit(xtr, ytr)
      pred = model.predict(xte)
      mse = mean_squared_error(pred,yte)

      
      #print('Mean squared error = {:5.2f}'.format(mean_squared_error(pred,yte)))
      elapsed_time = time.time() - start
      #print('Elapsed time = {:.4f} secs'.format(elapsed_time))

      if(best_mse < mse):
        best_mse = mse
        best_d = depth
        speed = elapsed_time
    print('Max_depth = {};, R_mse = {:7.4f}'.format(best_d , best_mse))
    print('Elapsed time = {:.4f} secs'.format(speed))
    print()



#best depth results are between 2 through 6
#depth of 4have best results overall
#for d in range(3,5)



print("randon splitter,rand_state = 42")
rand_DTR(x70_train, y70_train, x30_test, y30_test, rand_s)
rand_DTR(x80_train, y80_train, x20_test, y20_test, rand_s)
rand_DTR(x90_train, y90_train, x10_test, y10_test, rand_s)
print()

randon splitter,rand_state = 42
bestmf:  1
Max_depth = 1;, R_mse =  0.2384
Elapsed time = 0.0013 secs

bestmf:  2
Max_depth = 7;, R_mse =  0.2467
Elapsed time = 0.0010 secs

bestmf:  3
Max_depth = 7;, R_mse =  0.2467
Elapsed time = 0.0010 secs

bestmf:  4
Max_depth = 7;, R_mse =  0.2467
Elapsed time = 0.0010 secs

bestmf:  5
Max_depth = 7;, R_mse =  0.2467
Elapsed time = 0.0010 secs

bestmf:  6
Max_depth = 7;, R_mse =  0.2467
Elapsed time = 0.0010 secs

bestmf:  7
Max_depth = 7;, R_mse =  0.2467
Elapsed time = 0.0010 secs

bestmf:  8
Max_depth = 7;, R_mse =  0.2467
Elapsed time = 0.0010 secs

bestmf:  1
Max_depth = 1;, R_mse =  0.2378
Elapsed time = 0.0021 secs

bestmf:  2
Max_depth = 7;, R_mse =  0.2409
Elapsed time = 0.0010 secs

bestmf:  3
Max_depth = 5;, R_mse =  0.2444
Elapsed time = 0.0009 secs

bestmf:  4
Max_depth = 5;, R_mse =  0.2444
Elapsed time = 0.0009 secs

bestmf:  5
Max_depth = 5;, R_mse =  0.2444
Elapsed time = 0.0009 secs

bestmf:  6
Max_depth = 5;, R_mse =  0.2490
El

# Best Decision Tree Regressor

In [ ]:
def Best_DTR(xtr, ytr, xte, yte, d = 4):
  start = time.time()
  model = DecisionTreeRegressor( max_depth = d, random_state = 42, max_features=8).fit(xtr, ytr)
  pred = model.predict(xte)
  print('Mean squared error = {:5.2f}'.format(mean_squared_error(pred,yte)))

  elapsed_time = time.time() - start
  print('Elapsed time = {:.4f} secs'.format(elapsed_time))
Best_DTR(x70_train, y70_train, x30_test, y30_test)
Best_DTR(x80_train, y80_train, x20_test, y20_test)
Best_DTR(x90_train, y90_train, x10_test, y10_test)

Mean squared error =  0.23
Elapsed time = 0.0095 secs
Mean squared error =  0.22
Elapsed time = 0.0069 secs
Mean squared error =  0.24
Elapsed time = 0.0074 secs


# MSE and speed results of Decision Tree Regressor Model

best decision tree regressor model created was trained with the 80by20 dataset split, max_depth = 4, max_feature = 8, random_state =42, and using default splitter of 'best'. Random splitter hindered best mse results even with random_state being applied. The randomness that works well in this instance is applying random_state with max_features and in this dataset, max_features = 8 helped us reached the lowest mse. Mean squared error =  0.22
Elapsed time = 0.0122 secs



# **Best models for regressors with their paramaters** 

In [ ]:
#knn regressor bestt
print("Best knn Regressor")
#algorithm doesnt matter, and hits best at n = 9
auto_knnReg(x90_train, y90_train, x10_test, y10_test, 9)

print()
#best two regressor models
print("Bagging regressor using sklearn library bagging regressor")
best_dtrbag(x80_train, y80_train, x20_test, y20_test)

#if unable to use that library Bagging Regressor Library, 
# Best model with decision tree regressor with changing paramters
# depth = 4, and max_feature = 8, random_state = 42
print()
print("Best Decision Tree Regressor")
Best_DTR(x80_train, y80_train, x20_test, y20_test)

Best knn Regressor
Mean squared error =  0.25
Elapsed time = 0.1085 secs

Bagging regressor using sklearn library bagging regressor
Mean squared error =  0.22
Elapsed time = 0.1175 secs

Best Decision Tree Regressor
Mean squared error =  0.22
Elapsed time = 0.0066 secs


# **Conclusion! Best Models:**

## Best Classifers:

**Decision tree classifer**



with  max_depth = 9, 

random_state = 42, 

max_features = 8 

using entropy over gini,

and with a training set of 80% 
 
training and 20%test.

results:

Accuracy = 0.6898

Elapsed time = 0.0280 secs

## Best regressor :

**sklearn library bagging regressor**

 with random_state = 42, 

 and 80% training and 20%test

results:

Mean squared error =  0.21

Elapsed time = 0.1125 secs



**OR**
**decision Tree Regressor**

 with depth = 4, 
 
 and max_feature = 8, 
 
 random_state = 42, 
 
 80% training and 20%test.

 results:

 Mean squared error =  0.22 
 
 and Elapsed time = 0.0108 secs


# Part 2b

 (Mandatory for CS5361, extra credit for CS4361). The file water_potability_original.xlsx contains a superset of the data in water_potability.xlsx, but some of the instances have missing values, indicated as an empty cell in the excel file.

Repeat the experiments from part 2, but now use this bigger file. Replace missing values by the mean value of the corresponding attribute in the training set.

That is, if X_train[i,j] is missing, replace it by np.mean(X_train[ind,j]), where ind is the list of indices of examples where attribute j is not missing.

If X_test[i,j] is missing, replace it also by np.mean(X_train[ind,j]).

For extra credit, try a better way to 'guess' the missing values.

# **Steps on better way to guess missing values**

1.
2.
3.


In [ ]:
#upload the other xlsx file
from google.colab import files
uploaded = files.upload()

Saving water_potability_original.xlsx to water_potability_original.xlsx


In [ ]:
!ls #check both files are in colab before starting

sample_data


In [ ]:
import pandas as pd

water_df = pd.read_excel('water_potability_original.xlsx')

water_df

ph    Hardness  ...  Turbidity  Potability
0          NaN  204.890455  ...   2.963135           0
1     3.716080  129.422921  ...   4.500656           0
2     8.099124  224.236259  ...   3.055934           0
3     8.316766  214.373394  ...   4.628771           0
4     9.092223  181.101509  ...   4.075075           0
...        ...         ...  ...        ...         ...
3271  4.668102  193.681735  ...   4.435821           1
3272  7.808856  193.553212  ...   2.798243           1
3273  9.419510  175.762646  ...   3.298875           1
3274  5.126763  230.603758  ...   4.708658           1
3275  7.874671  195.102299  ...   2.309149           1

[3276 rows x 10 columns]

In [ ]:
water_df1 = pd.read_excel('water_potability.xlsx')

water_df1

ph    Hardness  ...  Turbidity  Potability
0     7.899452  210.734124  ...   2.853767           1
1     6.145148  197.541072  ...   4.600886           0
2     5.036454  190.164520  ...   3.534665           1
3     8.969697  195.744765  ...   3.983099           0
4     8.285072  151.573778  ...   4.274661           1
...        ...         ...  ...        ...         ...
2006  7.058866  191.548401  ...   3.937334           1
2007  8.281121  227.654351  ...   3.687995           0
2008  6.946718  214.170043  ...   4.934986           0
2009  7.783306  196.815952  ...   5.045620           0
2010  5.803497  193.200991  ...   2.181714           1

[2011 rows x 10 columns]

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import time
from sklearn.metrics import accuracy_score, confusion_matrix,mean_squared_error,mean_absolute_error
from sklearn.metrics import confusion_matrix

from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.naive_bayes import GaussianNB, MultinomialNB, ComplementNB, BernoulliNB

from sklearn.neighbors import KNeighborsRegressor
from sklearn.tree import DecisionTreeRegressor

from sklearn.ensemble import BaggingRegressor
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn import tree
from google.colab import files

water_df = pd.read_excel('water_potability.xlsx')

water_np = water_df.to_numpy() #(2011,10)


#split data into X dataset (water samples and their 9 attributes) 
#and y_values pair (potability found in last column)
X = water_np[:,0:9] #(2011, 9) 2-d array
y = water_np[:,9]   #(2011, 1) 1-d

# 80% by 20% row split = 1608, 403
# 90% by 10% row split = 1809, 202
# random_state=42, popular int random seed (controls randomness)

rs = 42  #Popular/common int value for random_state 0 and 42

#testing 3 sizes of data set splits in all algorithms
x70_train, x30_test, y70_train, y30_test = train_test_split(X, y, test_size=0.3, train_size = .7, random_state=rs)
x80_train, x20_test, y80_train, y20_test = train_test_split(X, y, test_size=0.2, train_size = .8, random_state=rs)
x90_train, x10_test, y90_train, y10_test = train_test_split(X, y, test_size=0.1, train_size = .9, random_state=rs)


FileNotFoundError: ignored

# **Part 3.** 
Write a report and submit it as an additional file in word or pdf format. Your report must include (at least) the following items:

Problem description
Algorithms implemented
Experimental results, including accuracies or mean squared errors and running times for each algorithm and parameter choice. Use tables and/or plots to illustrate this.
Discussion of results. How do algorithms compare in terms of accuracy/MSE and running times? What parameter choices work best? What parameter choices result in overfitting?
Conclusions. What did you learn from the assignment?

# **Report uploaded with lab on Word Doc** 